### PR TITLE
Implement --prune-whitelist(-w) flag to overwrite default whitelist for --prune

### LINKED ISSUE #35222: Need to support --dry-run for kubectl apply --prune
`--dry-run` is not acting properly with `kubectl apply --prune`, reproduction as below:

```
$ kubectl apply -f svc-nginx-1.yaml
service "my-nginx-1" created

$ kubectl apply -f svc-nginx-2.yaml --prune -l test=true --dry-run
service "my-nginx-2" created (dry run)
service "my-nginx-1" pruned
```

After the second command, service "my-nginx-1" actually got pruned. IIUC, `prune` should be also included by `dry-run`?

PS: "svc-nginx-1.yaml" and "svc-nginx-2.yaml" each contains one service with `test: "true"` label. Services' names are different.

### PR DESCRIPTION
**Pull request by MrHohn**
From kubernetes/kubectl#1854.

Updates:

As suggested, the new commits implement a default whitelist for `kubectl apply --prune`, which could be overwritten by using `--prune-whitelist`or `-w` flag. The default whitelist and example as below.

Also supports `--dry-run` for `--prune` to fix kubernetes/kubernetes#35222.

whitelist:

```
type pruneResource struct {
    group      string
    version    string
    kind       string
    namespaced bool
}

*pruneResources = []pruneResource{
    {"", "v1", "ConfigMap", true},
    {"", "v1", "Endpoints", true},
    {"", "v1", "Namespace", false},
    {"", "v1", "PersistentVolumeClaim", true},
    {"", "v1", "PersistentVolume", false},
    {"", "v1", "Pod", true},
    {"", "v1", "ReplicationController", true},
    {"", "v1", "Secret", true},
    {"", "v1", "Service", true},
    {"batch", "v1", "Job", true},
    {"extensions", "v1beta1", "DaemonSet", true},
    {"extensions", "v1beta1", "Deployment", true},
    {"extensions", "v1beta1", "HorizontalPodAutoscaler", true},
    {"extensions", "v1beta1", "Ingress", true},
    {"extensions", "v1beta1", "ReplicaSet", true},
    {"apps", "v1beta1", "StatefulSet", true},
}
```

example:

```
$ kubectl apply -f /path/to/file --prune -l test=true \
      -w core/v1/Pod \
      -w core/v1/Service \
      -w extensions/v1beta1/Deployment 
```

@mikedanese

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35220)

### CODE REVIEW

**Comment by mikedanese** [comment]
Thanks for picking this up. Would it make sense to just do all types always?

**Comment by MrHohn** [comment]
Both ways make sense to me. But people might have other concerns I am not aware of.
Open to advice.

**Comment by pwittrock** [comment]
I actually think we may want to solve this another way so it is more reproducible

**Comment by mikedanese** [comment]
> I actually think we may want to solve this another way so it is more reproducible

Can you elaborate or sync with @MrHohn on how to achieve this?

**Comment by pwittrock** [comment]
I need to come up with something more detailed.  The TLDR is that I think we want the options for kubectl apply to be stored in the directory that is applied.  `--prune`,  label selectors for matching resources, types of resources to prune, etc.  That way when you `apply` a directory the behavior isn't driven by the flags the user types in, but instead by the checked in configuration.

Config would
-  Specify an annotation applied to every resource it creates.  Only these resources would be deleted.  This would be used instead of the last-applied-config annotation.
- Prevent users from accidentally applying the wrong directory, or miss matching the directory and flags.
- Provide a way to manage how configuration is applied through checked in source instead of un reproducible commands.

May it is fine to skip this for now, but in that case --prune should only be used with CICD that ensures `apply` that is run is reproducible.

WDYT?

**Comment by mikedanese** [comment]
I think we need a solution for v1.5 that isn't a massive time sink for me or @MrHohn who have other commitments for v1.5. I think it makes sense to apply --prune for all types and not add a flag until the overhaul you are describing. This is an expert feature that provides real value to users who understand the caveats and we should document it as so until we can add some safety and reproducibility guarantees. Can we label apply --prune as a nonstable API (essentially alpha) for v1.5?

**Comment by pwittrock** [comment]
TLDR; That is fine lets document in the help that it is alpha and users must be prepared for backwards compatibility to break in subsequent releases.

I am not sure if we have a great way of communicating alpha features but yes we should definitely make this alpha.  Maybe in the help docs?  In 1.6 I would like --prune to fail if it cannot find a config.  Maybe in 1.6 prune should just require a path to the prune config or something.  Alternatively we could make it a subcommand instead so prune-specific flags.

**Comment by MrHohn** [comment]
Thanks for the discussions. I will soon change it to prune all types as a temporary solution.

**Comment by pwittrock** [comment]
WDYT of having a default whitelist that can be overridden with a flag?

**Comment by MrHohn** [comment]
Sounds reasonable. What about putting `core` and `extensions` groups into whitelist?

**Comment by MrHohn** [comment]
Though the `prune` feature are becoming stranger after this modification :(

Sorry for messing this up, but the addon manager really need this feature.

**Comment by MrHohn** [comment]
The new commits implement the default whitelist and the overridden flag. The detail is updated on the first post.

Currently the whitelist is hardcoded. I try to avoid hardcoding by using package like [the discovery client](https://github.com/kubernetes/kubernetes/tree/master/pkg/client/typed/discovery). But somehow the types return by server are also need to be manually filtered, some of them either do not have valid RESTMapping, or have RESTMapping but the server return error `could not find the resource`, or the server does not allow to list those resources. As it became more complex to distinguish those situations, I hardcoded the whitelist.

**Comment by MrHohn** [comment]
This is blocking #33698.

File: pkg/kubectl/cmd/apply.go [hunk 1]
```diff
@@ -123,6 +128,27 @@ func validatePruneAll(prune, all bool, selector string) error {
 	return nil
 }
 
+func validatePruneResources(pruneResources *[]pruneResource, gvks []string) error {
```
Comment (mikedanese) [inline pkg/kubectl/cmd/apply.go]: It's weird that this function is called `validatePruneResources` but it's actually modifying it's input argument. Is it doing to much or should it be renamed?

**Comment by mikedanese** [comment]
@fabianofranz @pwittrock PTAL. I'm pretty unfamiliar with kubectl so not sure if there is a better way to do this. This looks fine to me.

File: pkg/kubectl/cmd/apply.go (hunk 1 above)
Comment (MrHohn) [inline pkg/kubectl/cmd/apply.go]: Thanks for the comment. I changed the name to `parsePruneResources` and put it into RunApply().

**Comment by mikedanese** [comment]
@fabianofranz @pwittrock ping.

**Comment by pwittrock** [comment]
Generally looks good.  Comments are generally minor except for the StatefulSets pruning.

@mikedanese I will be OOO for the next 2 weeks with limited availability.  Would you be willing to close off any of the few comments I left?  Otherwise LGTM.

---

Reviewed 1 of 3 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/kubectl/cmd/apply.go, line 131 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KUYIsvD6cPHOBLnOVkg-r2-131:-KVVuvZ0IPGZRtq-Atfx:bbk5srr) ([raw file](https://github.com/kubernetes/kubernetes/blob/4b8676a13a7b576f6d13b78d05220e11d5276197/pkg/kubectl/cmd/apply.go#L131)):*
<details><summary><i>Previously, MrHohn (Zihong Zheng) wrote…</i></summary>
> Thanks for the comment. I changed the name to `parsePruneResources` and put it into RunApply().
</details>
Non-blocker: Why not just return a slice with the error instead of passing it in?

---

*[pkg/kubectl/cmd/apply.go, line 334 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KVVveeXD02helgwFH44:-KVVveeXD02helgwFH45:bu9dfu3) ([raw file](https://github.com/kubernetes/kubernetes/blob/858fa1a85c99c66234f67047d75e4bcf6dd35f0f/pkg/kubectl/cmd/apply.go#L334)):*
> ```Go
> 
> func (pr pruneResource) String() string {
> 	return fmt.Sprintf("%v/%v, Kind=%v, namespaced=%v", pr.group, pr.version, pr.kind, pr.namespaced)
> ```

nit: any particular reason Kind is capitalized while namespaced is not?

---

*[pkg/kubectl/cmd/apply.go, line 341 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KVVvA93YYGaqjl9o7wL:-KVVvA93YYGaqjl9o7wM:bw7r4ya) ([raw file](https://github.com/kubernetes/kubernetes/blob/858fa1a85c99c66234f67047d75e4bcf6dd35f0f/pkg/kubectl/cmd/apply.go#L341)):*
> ```Go
> 		// default whitelist
> 		*pruneResources = []pruneResource{
> 			{"", "v1", "ConfigMap", true},
> ```

What about StatefulSet?  Anything else new to beta or ga in 1.5?

---

*[pkg/kubectl/cmd/apply.go, line 350 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KVVvJFAHmEg2VYQTz4V:-KVVvJFAHmEg2VYQTz4W:b-d8zksg) ([raw file](https://github.com/kubernetes/kubernetes/blob/858fa1a85c99c66234f67047d75e4bcf6dd35f0f/pkg/kubectl/cmd/apply.go#L350)):*
> ```Go
> 			{"", "v1", "Secret", true},
> 			{"", "v1", "Service", true},
> 			{"batch", "v1", "Job", true},
> ```

What should we do with older api versions - e.g. v1beta1 jobs?  At least add a TODO about this and file an issue.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-:-KVVy-uPzrTDY5DNBdcQ:b-8f3ztk)*

**Comment by mikedanese** [comment]
@pwittrock Thanks for taking a look and giving feedback. I will be happy to finish the review.

**Comment by MrHohn** [comment]
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/kubectl/cmd/apply.go, line 131 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KUYIsvD6cPHOBLnOVkg-r2-131:-KVWG-UcGdO7HuXhqzjS:b10suy0) ([raw file](https://github.com/kubernetes/kubernetes/blob/4b8676a13a7b576f6d13b78d05220e11d5276197/pkg/kubectl/cmd/apply.go#L131)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> Non-blocker: Why not just return a slice with the error instead of passing it in?
</details>
Sure, changed.

---

*[pkg/kubectl/cmd/apply.go, line 334 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KVVveeXD02helgwFH44:-KVWGAlXNKw8ZpkkVWvD:bt2vr1b) ([raw file](https://github.com/kubernetes/kubernetes/blob/858fa1a85c99c66234f67047d75e4bcf6dd35f0f/pkg/kubectl/cmd/apply.go#L334)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> nit: any particular reason Kind is capitalized while namespaced is not?
</details>
Fixed.

---

*[pkg/kubectl/cmd/apply.go, line 341 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KVVvA93YYGaqjl9o7wL:-KVWGBshRBJJxA6G24G4:b-sxz49v) ([raw file](https://github.com/kubernetes/kubernetes/blob/858fa1a85c99c66234f67047d75e4bcf6dd35f0f/pkg/kubectl/cmd/apply.go#L341)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> What about StatefulSet?  Anything else new to beta or ga in 1.5?
</details>
Added StatefulSet.

---

*[pkg/kubectl/cmd/apply.go, line 350 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KVVvJFAHmEg2VYQTz4V:-KVWGF5ZZWhArslmesT9:bb3rkjp) ([raw file](https://github.com/kubernetes/kubernetes/blob/858fa1a85c99c66234f67047d75e4bcf6dd35f0f/pkg/kubectl/cmd/apply.go#L350)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> What should we do with older api versions - e.g. v1beta1 jobs?  At least add a TODO about this and file an issue.
</details>
Issue filed on #35991.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220)*

**Comment by MrHohn** [comment]
Many thanks for reviewing this! New commits are pushed.

Notice that `{"apps", "v1beta1", "StatefulSet", true}` is used for the StatefulSet, but PR #35731 that moves `StatuefulSet` from `v1alpha1` to `v1beta1` is umerged yet. However, it should not block this PR because for registeredMapper.RESTMapping() the version argument is just a preference but not mandatory.

In this case, RESTMapper will instead return `apps/v1alpha1, Kind=StatefulSet` if it does not have `v1beta1` yet. As below logs I got:
```
# use apps/v1beta1, Kind=StatefulSet
$ kubectl apply -f /a.yaml --prune -l prune-group=true --dry-run
pod "a" created (dry run)
...
[debug] apps/v1alpha1, Kind=StatefulSet
```

File: pkg/kubectl/cmd/apply.go [hunk 2]
```diff
@@ -271,26 +297,83 @@ func RunApply(f cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *App
...
+			{"extensions", "v1beta1", "Ingress", true},
+			{"extensions", "v1beta1", "ReplicaSet", true},
+			{"apps", "v1beta1", "StatefulSet", true},
+		}
+	}
+	registeredMapper := registered.RESTMapper()
+	for _, resource := range *pruneResources {
+		addedMapping, err := registeredMapper.RESTMapping(unversioned.GroupKind{Group: resource.group, Kind: resource.kind}, resource.version)
+		if err != nil {
+			return nil, nil, fmt.Errorf("invalid resrouce %v: %v", resource, err)
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/apply.go]: typo "resrouce"

File: pkg/kubectl/cmd/apply.go [hunk 3]
```diff
@@ -46,6 +48,7 @@ type ApplyOptions struct {
 	Prune           bool
 	Cascade         bool
 	GracePeriod     int
+	pruneResources  []pruneResource
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/apply.go]: Any reason why this is not exposed? Would make it better testable.

File: pkg/kubectl/cmd/apply.go [hunk 4]
```diff
@@ -101,6 +104,7 @@ func NewCmdApply(f cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmdutil.AddValidateFlags(cmd)
 	cmd.Flags().StringVarP(&options.Selector, "selector", "l", "", "Selector (label query) to filter on")
 	cmd.Flags().Bool("all", false, "[-all] to select all the specified resources.")
+	cmd.Flags().StringArrayP("groupversionkind", "g", []string{}, "Overwrite the default whitelist with <group/version/kind> for --prune")
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/apply.go]: We kind of try to have dashes in flag names as a convention. Given you have a short form, could this be `--group-version-kind`, or something directly related to the purpose instead of the format, like `--prune-whitelist`, `--skip-when-pruning`, `--prune-skipped`, or something similar? Starting with `--prune*` would also make them appear together in the flags list.  ;)

File: pkg/kubectl/cmd/apply.go [hunk 5]
```diff
@@ -135,6 +161,11 @@ func RunApply(f cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *App
 		return err
 	}
 
+	options.pruneResources, err = parsePruneResources(cmdutil.GetFlagStringArray(cmd, "groupversionkind"))
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/apply.go]: Is this flag mutually exclusive with another one, like `--all`? For example, what happens if you provide both `--all` and `--groupversionkind`? You may need to validate if they're mutually exclusive, or make it clear if one overrides the behavior of the other.

**Review by fabianofranz — COMMENTED** [review]
A few comments. Sorry about the delay!

File: pkg/kubectl/cmd/apply.go (hunk 4 above)
Comment (MrHohn) [inline pkg/kubectl/cmd/apply.go]: Make sense. Changing it to `--prune-whitelist`

File: pkg/kubectl/cmd/apply.go (hunk 5 above)
Comment (MrHohn) [inline pkg/kubectl/cmd/apply.go]: I think `--all` and `--gourpversionkind` should be allowed to appear together? As the former indicates all objects regardless to label selector and the latter indicates the resource type?

**Comment by MrHohn** [comment]
@k8s-bot unit test this, issue: #35872

**Comment by MrHohn** [comment]
From GCE e2e test:
```
fatal: unable to access 'https://github.com/kubernetes/release/': Could not resolve host: github.com
```

Seems unrelated.
@k8s-bot cvm gce e2e test this, issue: #IGNORE

**Comment by MrHohn** [comment]
The new commit changes `--groupversionkind`(`-g`) flag to `--prune-whitelist`(`-w`).

**Comment by MrHohn** [comment]
Codes were modified according to the instructions. Is it ready to go? @fabianofranz @mikedanese 

This is also blocking #33239.

**Comment by mikedanese** [comment]
LGTM

**Comment by MrHohn** [comment]
Many thanks!

**Comment by bprashanth** [comment]
@pwittrock I haven't been following this too closely, but regarding https://github.com/kubernetes/kubernetes/pull/35220#issuecomment-255236360

>TLDR; That is fine lets document in the help that it is alpha and users must be prepared for backwards compatibility to break in subsequent releases.

If you want it to truly be "alpha", I think we will have to flag control it, and disable it on platforms with a support contract (and think of another way to handle addon update by monday). 

I assume you mean this is effectively beta (i.e works on GKE).

**Comment by bprashanth** [comment]
Actually I don't think that matters. 
I don't even know what the guarantee is for client side alpha features against GKE clusters. If someone starts using it, and it works, it'll keep working till they upgrade the client. They're 100% in control, they just can't use 1.6 kubectl features AND have this feature.
